### PR TITLE
KMeans sparsity support

### PR DESCRIPTION
**Pull request by inteldimitrius**
# KMeans sparsity support
- [x] GPU kernel implemented
- [x] CPU DAAL kernel integrated
- [x] Distributed mode is supported
- [x] bazel tests + random blob dataset generation 
Implemented Lloyd method for sparse data.

### CODE REVIEW

**Review by Vika-F — COMMENTED** [review]
Thank you for adding sparse K-means implementation.
I have several comments:
- please replace `sparse` with `csr` in the names of the files to make them consistent with the rest of oneDAL.
- please add more descriptions in GPU implementation.
- I also think that cluster update step can be implemented more optimally. Seems that the range dimension over column_count is not needed in case of correct CSR format processing. I will try to provide the alternative implementation later.

Please see the other comments below.

File: cpp/oneapi/dal/algo/kmeans/backend/gpu/train_kernel_lloyd_sparse_dpc.cpp [hunk 1]
```diff
@@ -0,0 +1,662 @@
...
+                auto min_idx = max_index;
+                auto row_dists = distances_ptr + row_idx * cluster_count;
+                for (std::int32_t cluster_id = local_id; cluster_id < cluster_count;
+                    cluster_id += local_size) {
+                    const auto dist = cent_squares_ptr[cluster_id] + row_dists[cluster_id] +
+                                    data_squares_ptr[row_idx];
+                    if (dist < min_dist) {
+                        min_dist = dist;
+                        min_idx = cluster_id;
+                    }
```
[the remark this answers is not in the document]
Comment (inteldimitrius) [reply cpp/oneapi/dal/algo/kmeans/backend/gpu/train_kernel_lloyd_sparse_dpc.cpp]: No, it contains also distance computation & saving to global memory, so this part of code can't be replaced with reduction primitive.

File: cpp/oneapi/dal/algo/kmeans/detail/train_init_centroids.hpp [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+* limitations under the License.
+*******************************************************************************/
+#include <daal/src/algorithms/kmeans/kmeans_init_kernel.h>
+
+#include "oneapi/dal/backend/interop/common_dpc.hpp"
+#include "oneapi/dal/backend/interop/error_converter.hpp"
+#include "oneapi/dal/backend/interop/table_conversion.hpp"
+#include "oneapi/dal/backend/transfer.hpp"
+
+namespace oneapi::dal::kmeans::backend {
```
Comment (Vika-F) [inline cpp/oneapi/dal/algo/kmeans/detail/train_init_centroids.hpp:23]: Should it be `oneapi::dal::kmeans::detail`?

File: cpp/oneapi/dal/backend/primitives/reduction/reduction_dpc.cpp [hunk 3]
```diff
@@ -88,6 +88,50 @@ sycl::event reduce_by_rows_impl(sycl::queue& q,
     return sycl::event{};
 }
 
+template <typename Float, typename BinaryOp, typename UnaryOp>
+sycl::event reduce_by_rows_impl(sycl::queue& q,
```
Comment (Vika-F) [inline cpp/oneapi/dal/backend/primitives/reduction/reduction_dpc.cpp:92]: Please write the description and include the dimensions of the input arrays there as well.

**Review by Vika-F — COMMENTED** [review]
The changes look good to me. My last request is to improve the comments one more time.

**Review by Vika-F — APPROVED** [review]
LGTM after the tests pass.

**Pull request merged**